packages/common/nbstore/src/storage/storage.ts

```diff
@@ -9,6 +9,77 @@ export interface StorageOptions {
   id: string;
 }
 
+export function universalId({ peer, type, id }: StorageOptions) {
+  return `@peer(${peer});@type(${type});@id(${id});`;
+}
+
+export function isValidUniversalId(opts: Record<string, string>): boolean {
+  const requiredKeys: Array<keyof StorageOptions> = [
+    'peer',
+    'type',
+    'id',
+  ] as const;
+
+  for (const key of requiredKeys) {
+    if (!opts[key]) {
+      return false;
+    }
+  }
+
+  return opts.type === 'userspace' || opts.type === 'workspace';
+}
+
+export function parseUniversalId(id: string) {
```

Comment:
complex, why not split ';' then `/@(\w+)\((\w+)\)$/.match`

Reply:
regex is more complex in parsing and even slower

Reply:
<img width="924" alt="image" src="https://github.com/user-attachments/assets/8a7fd39b-5f57-4b03-a1f7-c5860b922ae9" />

speed not bad in this case

Reply:
but ';' will be treated as valid char in peer

Reply:
historical legacy issue?

Reply:
nope, but peer is supposed to be any string